server/container_create_linux.go

```diff
@@ -846,25 +846,52 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 		specgen.Config.Process.User.Umask = &umask
 	}
 
-	etc := filepath.Join(mountPoint, "/etc")
-	// create the `/etc` folder only when it doesn't exist
-	if _, err := os.Stat(etc); err != nil && os.IsNotExist(err) {
+	etcPath := filepath.Join(mountPoint, "/etc")
+
+	// Warn users if the container /etc directory path points to a location
+	// that is not a regular directory. This could indicate that something
+	// might be afoot.
+	etc, err := os.Lstat(etcPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, err
+	}
+	if err == nil && !etc.IsDir() {
+		log.Warnf(ctx, "Detected /etc path for container %s is not a directory", ctr.ID())
+	}
+
+	// The /etc directory can be subjected to various attempts on the path (directory)
+	// traversal attacks. As such, we need to ensure that its path will be relative to
+	// the base (or root, if you wish) of the container to mitigate a container escape.
+	etcPath, err = securejoin.SecureJoin(mountPoint, "/etc")
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve container /etc directory path: %w", err)
+	}
+
+	// Create the /etc directory only when it doesn't exist.
+	if _, err := os.Stat(etcPath); err != nil && os.IsNotExist(err) {
 		rootPair := idtools.IDPair{UID: 0, GID: 0}
 		if containerIDMappings != nil {
 			rootPair = containerIDMappings.RootPair()
 		}
-		if err := idtools.MkdirAllAndChown(etc, 0o755, rootPair); err != nil {
-			return nil, fmt.Errorf("error creating mtab directory: %w", err)
+		if err := idtools.MkdirAllAndChown(etcPath, 0o755, rootPair); err != nil {
+			return nil, fmt.Errorf("failed to create container /etc directory: %w", err)
 		}
 	}
-	// add symlink /etc/mtab to /proc/mounts allow looking for mountfiles there in the container
-	// compatible with Docker
-	if err := os.Symlink("/proc/mounts", filepath.Join(etc, "mtab")); err != nil && !os.IsExist(err) {
-		return nil, err
+
+	// we cannot use SecureJoin here, as /etc/mtab is already being symlinked from somewhere in some cases
+	// and doing so resolves mtabPath to mountPoint/etc/proc/self/mounts, which breaks container creation.
+	// Let's just check if mtab is created already, and if not, symlink it.
+	mtabPath := filepath.Join(etcPath, "mtab")
+	if _, err := os.Stat(mtabPath); err != nil && os.IsNotExist(err) {
+		// Add symbolic link from /proc/mounts to /etc/mtab to keep compatibility with
+		// legacy Linux distributions and Docker.
+		if err := os.Symlink("/proc/mounts", mtabPath); err != nil && !os.IsExist(err) {
```

Comment:
We can probably now drop the following:

```suggestion
		if err := os.Symlink("/proc/mounts", mtabPath); err != nil {
```

Since we only add the symbolic link now when it does not exist.

Reply:
Oh, I first misread that line. I missed the `!`. I think this would make it even more explicit:

```
if err := os.Symlink("/proc/mounts", mtabPath); err != nil && os.IsExist(err) 
```

Reply:
good catch